### ISSUE TITLE
Channels Tool seems broken on MacOS FIJI/ImageJ

### ISSUE THREAD

**Issue by hyungsongnam** [imagej/imagej#240]
Hello,

I recently noticed that the "Channels Tool" seems to be broken when running FIJI/ImageJ on mac OS with Java 8.

The behavior is as follows:
(1) open an image
(2) open channels tool - Can change from the dropdown menu from composite, color, etc
(3) zoom in zoom out - Now, the dropdown menu in the channels tool becomes unresponsive - it cannot be changed.

If I quit and restart ImageJ, the channels tool works again, but only on the first image - if I close the first image and open another image, the tool becomes unresponsive again.

When running the FIJI/ImageJ on Ubuntu Linux, and opening the same image files, I don't see the same behavior with the channels tool, so this behavior might be mac OS specific.

Thank you.

Sincerely,

Hyung-song Nam, PhD

**Comment by ctrueden** [comment]
Thanks for the report, @hyungsongnam. I was able to reproduce this issue on my system with ImageJ 1.52o and 1.52p. I believe the patch which causes the issue is here:

https://github.com/imagej/ImageJA/commit/7ecb3e177e30744cb0b3e65ce9d877f7ec702e7c#diff-6eeb9bb47ee5d78fdba73066706c99cdL49

@rasband It seems dangerous to me to force the ImageJ main window to have the focus while the user is trying to manipulate a plugin dialog. Are you able to reproduce the issue? If not, I can grab a video and post it here, to illustrate what goes wrong.

**Comment by rasband** [comment]
I was not able to reproduce this problem using the Mitosis, Organ of Corti and Neuron sample images. The menu bar would disappear if ImageJ did not force the main window to have the focus.

**Comment by hyungsongnam** [comment]
I have looked for this behavior with additional images.
I can add that this is sporadic. 
However, it does occur, and I have attached a movie from my mac.
Sorry for the long movie, but the issue I'm describing occurs at the end of the movie.
Thank you.
Sincerely,
Hyung-song Nam, PhD


[Hyung-song's FIJI-ImageJ movie.zip](https://github.com/imagej/imagej/files/3528298/Hyung-song.s.FIJI-ImageJ.movie.zip)

**Comment by hyungsongnam** [comment]
I've done more testing of this.
It is indeed sporadic, and finding a reproducible trigger was difficult.
The behavior appears to be associated with changing focus of the windows.
Thank you.
Sincerely,
Hyung-song Nam, PhD

**Comment by rasband** [comment]
I was not able to reproduce this problem but it should be fixed in the latest ImageJ daily build (1.52q42). To upgrade, use the Help>Update ImageJ command and select "daily build" from the drop down menu. ImageJ no longer forces the main window to have the focus when activating the Channels window on macOS. As a result, the menu bar disappears when you switch from an image window to the Channels window.

**Comment by hyungsongnam** [comment]
Thank you.

Sincerely,
Hyung-song Nam, PhD

**Comment by hyungsongnam** [comment]
In an attempt to reconcile the different findings, I re-installed the no-jre version of FIJI downloaded today, and repeated the testing.
Strangely, I cannot reproduce the behavior at all today, even after trying for some time.
Since there is a video of the behavior from yesterday, I observed it yesterday.
Maybe my installation of FIJI simply needed an update.
I can't say for sure whether today's change to the code will be necessary, the developers can decide.

Thank you.

Sincerely,
Hyung-song Nam, PhD

**Comment by rasband** [comment]
The problem my be related to Java. What version of Java are you using now and what version were you using when you were having the problem? I am using Java 1.8.0_172 and I do not see the problem.

**Comment by ctrueden** [comment]
I used AdoptOpenJDK 1.8.0_202 in my tests, where I did see the problem.

The reason I called stealing the focus from the Channels window "dangerous" is because it goes against typical windowing assumptions: the user manipulates the active window, not background windows. There may be environments where manipulating an inactive window with the mouse does not work at all—as we've seen, in at least some macOS-based environments it is not guaranteed to work. It also precludes keyboard-based manipulation of the dialog. Therefore, I support the decision to keep the dialog in focus, even if the menu bar must disappear as a consequence.

But I was wondering: why does the menu disappear for the Channels tool? The `PlugInDialog` constructor uses `IJ.getInstance()` as the parent frame on macOS:

https://github.com/imagej/ImageJA/blob/v1.52p/src/main/java/ij/plugin/frame/PlugInDialog.java#L13-L14

This should preserve the menu bar from the main window, no? It works in a test script that I wrote:

<details><summary>Menu_Test.groovy</summary>

```groovy
import java.awt.*
import javax.swing.*

frame = new JFrame("Test Frame")
menubar = new JMenuBar()
file = new JMenu("File")
menubar.add(file)
file.add(new JMenuItem("Test"))
frame.setJMenuBar(menubar)
frame.setBounds(100, 100, 400, 300)
frame.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE)
frame.setVisible(true)

//dialog = new JDialog(frame)
//dialog = new JDialog(ij.IJ.getInstance())
dialog = new JDialog(null)
dialog.getContentPane().add(new JButton("Test"))
dialog.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE)
dialog.setModal(false)
dialog.pack()
dialog.setVisible(true)
```

</details>

I'll dig a little more right now to see if I can narrow down the issue.

**Comment by ctrueden** [comment]
@rasband Here is what I have learned so far:

1. The `windowActivated` method of `ImageWindow` steals the main ImageJ window's menu bar ([1](https://github.com/imagej/ImageJA/blob/v1.52p/src/main/java/ij/gui/ImageWindow.java#L554-L555), [2](https://github.com/imagej/ImageJA/blob/v1.52p/src/main/java/ij/gui/ImageWindow.java#L704)).
2. Internally, when the `ImageWindow` sets its `MenuBar` to ImageJ's menu bar, that menu bar is removed from its previous parent, which in this case is the main ImageJ window ([1](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/43cb25339b55/src/share/classes/java/awt/Frame.java#l581)).
3. To combat this wrinkle, the main ImageJ window grabs the menu bar again any time it is reactivated ([1](https://github.com/imagej/ImageJA/blob/v1.52p/src/main/java/ij/ImageJ.java#L642-L652)). In general, whenever a `Frame` is activated on macOS, it assigns the menu bar to itself, which ensures that from the user's perspective (apart from a brief flash), the menu bar remains available.
4. Unfortunately, when a `Dialog` is activated, this same trick does not work: resetting the menu bar of the main ImageJ window from within the `windowActivated` method of `PlugInDialog` does not cause the dialog's menu bar to successfully appear.
5. I also tried resetting the main ImageJ window's menu bar from the `windowDeactivated` method of `ImageWindow`, hoping that that would be early enough for the dialog activation to pick up its parent's menu bar successfully. But alas not.

Now I am looking for some other invocation that could be used from the `windowActivated` method of `PlugInDialog` to force the screen menu bar to refresh/repaint.

**Comment by ctrueden** [comment]
Unfortunately, I am out of time, and did not find a solution yet. One idea I had was to "bounce back"—i.e. activate the main window and immediately reactivate the dialog, to refresh the menu bar—but I think that approach is very hacky and my intuition is that it could create more bugs than it solves.

My next idea was/is to dig into the Apple-specific Java code (e.g. [here](http://hg.openjdk.java.net/macosx-port/macosx-port/jdk)) and try to understand it a bit better, to see if there is an API hook (even a private one) that could be called to force a screen menu bar refresh.

**Comment by rasband** [comment]
Hi Curtis,

In the latest ImageJ daily build (1.52q43), I changed the code in PlugInDialog that restores the Mac menu bar to
```
   if (IJ.isMacOSX() && ij!=null) {
      ij.requestFocus();
      this.requestFocus();
   }

```
Does this also fix the problem you saw with the Channels dialog on AdoptOpenJDK 1.8.0_202?

**Comment by rasband** [comment]
This 1.52o regression is fixed in the ImageJ 1.52q44 daily build.
